python/ray/dag/dag_node.py

```diff
@@ -344,28 +345,56 @@ def _apply_and_replace_all_child_nodes(
         Returns:
             New DAGNode after replacing all child nodes.
         """
-
         replace_table = {}
         # CloudPickler scanner object for current layer of DAGNode. Same
         # scanner should be use for a full find & replace cycle.
         scanner = _PyObjScanner()
+
+        # Dependency graph for detecting and enforcing control dependencies
+        # based on bind_index in addition to data dependencies
+        dep_graph = DependencyGraph()
+
         # Find all first-level nested DAGNode children in args.
-        # Update replacement table and execute the replace.
-        for node in scanner.find_nodes(
+        all_nodes = scanner.find_nodes(
             [
                 self._bound_args,
                 self._bound_kwargs,
                 self._bound_other_args_to_resolve,
             ]
-        ):
-            if node not in replace_table:
+        )
+
+        # Create a map of nodes to their bind_index if present
+        bind_indices = {
+            node: node.get_other_args_to_resolve().get("bind_index")
+            for node in all_nodes
+            if "bind_index" in node.get_other_args_to_resolve()
+        }
+
+        # Iterate through each pair of nodes and add control dependencies
+        for node, node_bind_index in bind_indices.items():
+            for other_node, other_bind_index in bind_indices.items():
+                if other_bind_index < node_bind_index:
+                    dep_graph.add_dependency(other_node, node)
+
+        # Topologically sort the control dependencies for this layer of the dag
+        ctrl_dependencies = dep_graph.topological_sort()
```

Comment:
I believe the nodes here are only for the current DAG node's arguments, not the entire DAG? In that case, I'm not sure that topological sort of just the current args will necessarily work.

```python
with InputNode() as inp:
  x1 = a.foo.bind(inp)
  x2 = a.bar.bind(inp)
  y1 = b.baz.bind(x1)
  y2 = c.baz.bind(x2)
  dag = MultiOutputNode([y2, y1])
```
In this case, y2 and y1 execute on different actors and they have the same bind index, so we may visit y2 then y1. Then since it does DFS, it should execute x2, y2, x1, y1. But we want to make sure to execute x1 before x2.

Topological sort can work, but you would need to modify how the recursion is done to not do DFS I believe. The simpler way is probably to add a control dependency during DAG declaration instead of execution time, by adding the previous DAG node submitted to the same actor to the nodes found by `scanner.find_nodes`.